hicommonwealth/commonwealth
# Fix/Clean Discobot

## Link to Issue
Closes: #5462 
Closes: #5515
Closes: #5519
Closes: #5548 
Closes: #5563 

## Description of Changes
- Added discord-bot type check to CI
- Typed the `discord_meta` column on `Threads` and `Comments` table
- Refactored the Discord consumer and listener for clarity and bug fixes:
  - Used a handler pattern
  - Update actions so they are more specific and message path/route can be easily determined
  - Extracted and updated the topic query into a separate function used by both the consumer and listener
- Added rollbar error reporting to the Discord bot

## Test Plan
- Follow the steps under **Testing** in the `Discobot.md` doc updated in this PR.

Between each step below you should ensure that the Discord action is reflected on the Commonwealth forum by  refreshing.
1. Create a post on Discord
2. Comment on the post
3. Edit the comment
4. Edit the post title
5. Delete the comment
6. Delete the post

## Deployment Plan

1. Turn off the Discobot consumer
2.. Deploy to Discobot and CW
3. Once both are deployed, check RabbitMQ Discord queue for messages if it is empty, re-enable the discobot consumer. If the queue is not empty, manually republish the messages from the RabbitMQ management dashboard with the correct `action` property defined.
4. Add `ROLLBAR_ENV` to the discobot environment variables. Set to `discobot`.
5. Add Datadog component for `cw.discobot_message_processed` `StatsD `COUNT` metric. Add Datadog component for the Discobot RabbitMQ queue.
6. Delete `DISCORD_UI_URL`, `DISCORD_OAUTH_SCOPES`, `DISCORD_OAUTH_CALLBACK`, `DISCORD_BOT_URL`, and `DISCORD_BOT_SUCCESS_URL` from production environment variables
7. While we are at it, also remvoe `CE_DATABASE_URL` and `CHAIN_EVENT_SERVICE_SECRET` which are no longer needed since chain-events was deprecated and Substrate removal was merged.

## Other Considerations

- The Discobot consumer can only be tied to a single database and app so it isn't possible to use the `discobot-listener-staging` app for all of our staging environments (frick, frack, beta). I think given that this PR clarifies how to run the bot locally, the `discobot-listener-staging` app should be tied to the `beta` environment so that @ilijabojanovic can tie the Discobot into his QA process. All other environments will not be able to test Discobot unless we create additional Discobot staging apps. Given that changes to the Discobot are quite rare, local testing + QA testing should be sufficient (frick/frack testing unnecessary).
